R/manual_layer_dag.R

```diff
@@ -0,0 +1,192 @@
+# DAG is an R reference class. We need references since we have a mutable DAG of nodes
+# and we must be able to update them as their computation progresses.
+#
+# Nominally this is invoked by the terminal node's compute() method.
+dagGenerator <- setRefClass("DAG", representation(
+  namespace         = "character",
+  all_nodes         = "list",
+  initial_nodes     = "list",
+  terminal_nodes    = "list"
+))
+
+dagGenerator$methods(
+
+  # ================================================================
+  # INITIALIZATION
+
+  initialize = function(.self, terminal_node, namespace) {
+    .self$namespace <- namespace
+    # TODO: pluralize everywhere, so we can have multiple terminal nodes.
+    .self$populate(terminal_node)
+  },
+
+  # Nodes already have a graph structure. For example:
+  #   a <- delayed(function()    {    9  },                    display_name='a', do_local=FALSE)
+  #   b <- delayed(function(x)   {  10*x },    args=list(a),   display_name='b', do_local=FALSE)
+  #   c <- delayed(function(x)   { 100*x },    args=list(a),   display_name='c', do_local=FALSE)
+  #   d <- delayed(function(...) { sum(...) }, args=list(b,c), display_name='d', do_local=FALSE)
+  # has DAG like
+  #    a
+  #   / \
+  #  v   v
+  #  b   c
+  #   \ /
+  #    v
+  #    d
+  #
+  # What we do here is locate the initial and terminal nodes.
+  populate = function(.self, terminal_node) {
+    if (!is(terminal_node, "Node")) {
+      stop("Terminal node must be an instance of the Node class; got ", class(terminal_node))
+    }
+    .self$all_nodes <- list()
+    .self$initial_nodes <- list()
+    .self$populate_aux(terminal_node)
+    if (length(.self$all_nodes) == 0) {
+      stop("Task graph is empty")
+    }
+    if (length(.self$initial_nodes) == 0) {
+      stop("Task graph has a cycle")
+    }
+
+    .self$terminal_nodes <- list(terminal_node)
+
+    # This makes our print output non-frustrating for the user to read
+    .self$sort_nodes_topologically()
+  },
+
+  # This is a recursive helper function for populate().
+  populate_aux = function(.self, node) {
+    if (!is.null(.self$all_nodes[[node$id]])) {
+      # In R, just "return" does not actually ... return. ;)
+      return()
+    }
+
+    .self$all_nodes[[node$id]] <- node
+
+    if (node$is_initial()) {
+      .self$initial_nodes[[node$id]] <- node
+    }
+
+    for (arg in node$args) {
+      if (is(arg, "Node")) {
+        .self$populate_aux(arg)
+      }
+    }
+  },
+
+  # This is important so our status display is non-frustrating for the user.
+  # This assumes a cyclicity check has already been done.
+  # https://en.wikipedia.org/wiki/Topological_sorting Technically this is a
+  # reverse topological sort -- terminals appear first, then the ones that
+  # depend on them. This is intuitive for the user (and this topological sort
```

Comment:
This will actually come in handy if we choose to implement task graph logging for R; the graphs must be topologically sorted when submitted for logging.